lib/oli_web/live/admin/external_tools/new_external_tool_view.ex

```diff
@@ -0,0 +1,256 @@
+          required
+        />
+        <.input
+          class="form-control h-11 placeholder:pl-6"
+          field={f[:keyset_url]}
+          type="text"
+          label="Keyset URL"
+          label_class="mb-2"
+          placeholder="Type here..."
+          additional_text={~H'<span class="text-red-500">(*Required)</span>'}
+          required
+        />
+        <.input
+          class="form-control mt-2 placeholder:pl-6"
+          group_class=""
+          field={f[:redirect_uris]}
+          type="textarea"
+          data-grow="true"
+          autocomplete="off"
+          rows="3"
+          label="Redirect URIs"
+          label_class="mb-2"
+          placeholder="Type here..."
+          additional_text={~H'<span class="text-red-500">(*Required)</span>'}
+          required
+        />
+        <.input
+          class="form-control mt-2 placeholder:pl-6"
+          group_class=""
+          field={f[:custom_params]}
+          type="textarea"
+          data-grow="true"
+          autocomplete="off"
+          rows="3"
+          label="Custom Params"
+          label_class="mb-2"
+          placeholder="Type here..."
+        />
+
+        <div class="flex justify-end gap-2 my-8">
+          <.button
+            href={~p"/admin/external_tools"}
+            class="px-6 py-2 bg-white text-[#006cd9] border border-blue-500 rounded-md"
+          >
+            Cancel
+          </.button>
+          <.button
+            type="submit"
+            class="px-6 py-2 bg-[#0062F2] hover:bg-[#0075EB] text-white rounded-md"
+          >
+            Add Tool
+          </.button>
+        </div>
+      </.form>
+    </div>
+    """
+  end
+
+  def handle_event("validate", %{"tool_form" => params}, socket) do
+    changeset =
+      %PlatformInstance{}
+      |> PlatformInstances.change_platform_instance(params)
+      |> Map.put(:action, :validate)
+
+    {:noreply,
+     socket
+     |> assign(form: to_form(changeset, as: :tool_form))
+     |> assign(:custom_flash, nil)}
+  end
+
+  def handle_event("create_tool", %{"tool_form" => params}, socket) do
+    try do
+      case PlatformExternalTools.register_lti_external_tool_activity(params) do
+        {:ok, _tool} ->
+          new_changeset = PlatformInstances.change_platform_instance(%PlatformInstance{})
+
+          {:noreply,
+           socket
+           |> assign(form: to_form(new_changeset, as: :tool_form))
+           |> assign(:custom_flash, %{
+             type: :success,
+             message: "You have successfully added an LTI 1.3 External Tool at the system level."
+           })}
+
+        {:error, %Ecto.Changeset{} = changeset} ->
+          {:noreply,
+           socket
+           |> assign(form: to_form(changeset, as: :tool_form))
+           |> assign(:custom_flash, %{
+             type: :error,
+             message: "One or more of the required fields is missing; please check your input."
+           })}
+      end
+    rescue
```

Comment:
Couldn't we leverage changeset validations and avoid using a `try rescue` block here? This works, but does not seem to be the standard way of handling database errors such as unique constraints. Since this is a schema from a library, one way could be to extend or redefine the existing `PlatformInstance` changeset to have a `unique_constraint` over that field. Another option could be to define the `register_lti_external_tool_activity` function as an `Ecto.Multi` instead of a `Repo.transaction` and validate there the uniqueness in the pipeline.